storage/engine: remove InMem type

petermattis opened the pull request:
The InMem type was specific to RocksDB and only used in one place
outside of the storage/engine package. This is prep-work for allowing
`engine.NewInMem` return either a `*RocksDB` or `*Pebble`.

See #41987

Release note: None

petermattis committed:
storage/engine: remove InMem type

The InMem type was specific to RocksDB and only used in one place
outside of the storage/engine package. This is prep-work for allowing
`engine.NewInMem` return either a `*RocksDB` or `*Pebble`.

See #41987

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42017)
<!-- Reviewable:end -->

petermattis committed:
storage/engine: allow test servers to use in-memory pebble engines

Add an engine type parameter to `engine.NewInMem()` and plumb through
setting when test servers start. This causes test servers and test
clusters to respect `COCKROACH_TEST_STORAGE_ENGINE` when configured with
in-memory engines.

Replace most users of `engine.NewInMem()` with
`engine.NewDefaultInMem()` which selects either RocksDB or Pebble
depending on the setting of `COCKROACH_TEST_STORAGE_ENGINE`.

See #41987

Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42017#-:-LsSJCZt1QNLwrI9AZB3:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, @petermattis, and @sumeerbhola)

___
*[pkg/storage/replica_proposal.go, line 493 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42017#-LsSIukp1NgZH4vswVtW:-LsSIukp1NgZH4vswVtX:b-8v7alx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f5447d2739afeab12b8b9f616acc5e8501cad03/pkg/storage/replica_proposal.go#L493)):*
> ```Go
> 		}
> 	} else {
> 		ingestPath := path + ".ingested"
> ```

the commentary below and in `IngestExternalFiles` and `DBIngestExternalFiles` is rather complicated. Are all our customers running on recent enough RocksDB that all of this could be simplified?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42017#-:-LsSJ_Zg7Yh99xr3wmxy:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr, @itsbilal, and @sumeerbhola)

___
*[pkg/storage/replica_proposal.go, line 493 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42017#-LsSIukp1NgZH4vswVtW:-LsSJLDd6WT7eoJ4E8ff:bm3tefc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f5447d2739afeab12b8b9f616acc5e8501cad03/pkg/storage/replica_proposal.go#L493)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

the commentary below and in `IngestExternalFiles` and `DBIngestExternalFiles` is rather complicated. Are all our customers running on recent enough RocksDB that all of this could be simplified?
</blockquote></details>

I was looking at this as well. I think we're on a recent enough version of RocksDB. I worry about how this works on Windows. Certainly something to keep in mind for the future. I'll file an issue so we don't lose track.


<!-- Sent from Reviewable.io -->
